src/main/java/antlr/TreepatVisitorImplementationFunctions.kt

```diff
@@ -0,0 +1,109 @@
+package antlr
+
+import antlr.generate.TreepatParser
+import antlr.generate.TreepatVisitor
+import functions.ChildFunction
+import functions.NodeFunction
+import functions.SiblingFunction
+import org.antlr.v4.runtime.tree.ErrorNode
+import org.antlr.v4.runtime.tree.ParseTree
+import org.antlr.v4.runtime.tree.RuleNode
+import org.antlr.v4.runtime.tree.TerminalNode
+import tree.TargetTreeNode
+import java.util.stream.Collectors
+
+class TreepatVisitorImplementationFunctions : TreepatVisitor<(TargetTreeNode) -> List<TargetTreeNode>> {
+    override fun visitModel(ctx: TreepatParser.ModelContext?): (TargetTreeNode) -> List<TargetTreeNode> {
+        return ctx!!.subtree().accept<(TargetTreeNode) -> List<TargetTreeNode>>(this)
+    }
+
+    override fun visitSubtree(ctx: TreepatParser.SubtreeContext?): (TargetTreeNode) -> List<TargetTreeNode> {
+        val expression = ctx!!.expression().accept<(TargetTreeNode) -> List<TargetTreeNode>>(this)
```

Comment:
I think we can do something like:
```
val expression = ctx!!.expression().accept<VisitorFunction>(this)
val child = ctx.child()?.accept<VisitorFunction>(this) ?: null
return child ?: expression
``` 

I have not checked if the snippet above works but the idea is avoid `if(obj == null)` checks and try to take advantage of nullable types of Kotlin. 😊 

If above idea works maybe we could use the same pattern for subsequent methods if it's possible.

Reply:
Yes this a good idea but need to check the logic 🤔